PSPs/psp-22.md

```diff
@@ -9,181 +9,674 @@
+  ],
+  "mutates": true,
+  "name": [
+    "PSP22",
+    "decrease_allowance"
+  ],
+  "payable": false,
+  "returnType": null,
+  "selector": "0x1f86d882"
+}
+```
 
-‼️ Important ‼️
+#### PSP22Metadata 
+PSP22Metadata is an optional interface of metadata for Fungible Token Standard
+
+##### **token_name**() -> Option<String>
+```json
+{
+  "args": [],
+  "docs": [
+    " Returns the token name."
+  ],
+  "mutates": false,
+  "name": [
+    "PSP22Metadata",
+    "token_name"
+  ],
+  "payable": false,
+  "returnType": {
+    "displayName": [
+      "Option"
+    ],
+    "type": "Option<str>"
+  },
+  "selector": "0x9c994fe4"
+}
+```
 
-Events are not supported currently due to how ink! currently handles them.  
-The identifiers of events must be based on the name of the trait. At the moment, ink! doesn't support it,
-but it must be fixed with this [issue](https://github.com/paritytech/ink/issues/809). 
+##### **token_symbol**() -> Option<String>
+```json
+{
+  "args": [],
+  "docs": [
+    " Returns the token symbol."
+  ],
+  "mutates": false,
+  "name": [
+    "PSP22Metadata",
+    "token_symbol"
+  ],
+  "payable": false,
+  "returnType": {
+    "displayName": [
+      "Option"
+    ],
+    "type": "Option<str>"
+  },
+  "selector": "0x10972330"
+}
+```
 
-#### Transfer 
-Must be emitted when a token transfer occurs.
+##### **token_decimals**() -> u8
+```json
+{
+  "args": [],
+  "docs": [
+    " Returns the token decimals."
+  ],
+  "mutates": false,
+  "name": [
+    "PSP22Metadata",
+    "token_decimals"
+  ],
+  "payable": false,
+  "returnType": {
+    "displayName": [
+      "u8"
+    ],
+    "type": "u8"
+  },
+  "selector": "0x997ad16c"
+}
+```
+
+#### PSP22Receiver
+PSP22Receiver is an interface for any contract that wants to support safe transfers from PSP22 token smart contracts to avoid unexpected tokens on balance of contract.
+This method is called before transfer to ensure the recipient of the tokens acknowledge receipt
+
+##### **before_received**(&mut self, operator: AccountId, from: AccountId, value: Balance, data: Vec<u8>) -> Result<(), PSP22ReceiverError>
+```json
+{
+  "args": [
+    {
+      "name": "operator",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "name": "from",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "name": "value",
+      "type": {
+        "displayName": [
+          "Balance"
+        ],
+        "type": "Balance"
+      }
+    },
+    {
+      "name": "data",
+      "type": {
+        "displayName": [
+          "Vec"
+        ],
+        "type": "Vec<u8>"
+      }
+    }
+  ],
+  "docs": [
+    " Ensure that the smart contract allow reception of PSP22 token.",
+    " Returns `Ok(())` if the contract allow the reception of the token(s) and Error `TransferRejected(String))` otherwise.",
```

Comment:
```suggestion
    " Ensures that the smart contract allows reception of PSP22 tokens.",
    " Returns `Ok(())` if the contract allows the reception of the token(s) and Error `TransferRejected(String))` otherwise.",
```